### ISSUE TITLE
Diff Function does not update Operation

### ISSUE THREAD

**Issue by FUUbi** [yidafu/x-tree-diff-plus#3]
Hello :) Thanks for making the Source Code available. 

Apparently  `diff()` does not update  `tree.OP ` correctly.  If the old tree contains two text nodes with the same `value`, the function updates the `OP` correctly, but if they are distinct, only the hashes are updated and the `OP` remains unchanged. I reproduced this behaviour with the following code.
```typescript

    const getLog = (
        tree : XTree ,
        level = 1
    ) : string => {
        const space = Array ( level * 4 )
            .fill ( ' ' )
            .reduce ( (
                          a ,
                          b
                      ) => a + b )
        let log     = ''
        log += `${ space }Operation: ${ tree.Op }\n`
        log += `${ space }Label: ${ tree.label }\n`
        log += `${ space }Type: ${ tree.type }\n`
        log += `${ space }Value: ${ tree.value }\n`
        log += `${ space }Index: ${ tree.index }\n`
        log += `${ space }nMD: ${ tree.nMD }\n`
        log += `${ space }tMD: ${ tree.tMD }\n\n`


        if ( tree.hasChildren () ) {
            level++
            tree.forEach ( c => {
                const nextLog = getLog (
                    c ,
                    level )
                log += nextLog
            } )
        }


        return log
    }

    const DIV = ( children : XTree[] = [] ) => new XTree (
        {
            type :  NodeType.ELEMENT ,
            index : 1 ,
            label : 'DIV' ,
            children
        } )

    const H1 = (
        children : XTree[] = [] ,
        index              = 1
    ) => new XTree (
        {
            type :  NodeType.ELEMENT ,
            index ,
            label : 'H1' ,
            children
        } )

    const hey = ( index = 1 ) => new XTree (
        {
            type :  NodeType.TEXT ,
            index ,
            value : 'hey!'
        }
    )

    const well_done = ( index = 1 ) => new XTree (
        {
            type :  NodeType.TEXT ,
            index ,
            value : 'well done!'
        }
    )

    const great_work = ( index = 1 ) => new XTree (
        {
            type :  NodeType.TEXT ,
            index ,
            value : 'great work!'
        }
    )

    //           <div>
    //                  <h1>hey!</h1>
    //                  <h1>hey!</h1>
    //           </div>
    let tree1 = DIV ( [
                          H1 (
                              [ hey () ] ,
                              1
                          ) ,
                          H1 (
                              [ hey () ] ,
                              2
                          ) ,
                      ] )

    //           <div>
    //                  <h1>hey!</h1>
    //                  <h1>well done!</h1>
    //           </div>
    let tree2 = DIV ( [
                          H1 (
                              [ hey () ] ,
                              1
                          ) ,
                          H1 (
                              [ well_done () ] ,
                              2
                          ) ,
                      ] )


    new DefaultXTreeDiff (
        tree1 ,
        tree2
    ).diff ()
    console.log ( getLog ( tree1 ) )
    console.log ( getLog ( tree2 ) )

    /*

    TREE 1                                            TREE 2
    ---------------------------------------           ---------------------------------------
    Operation: 0                                      Operation: 0
    Label: DIV                                          Label: DIV
    Type: ELEMENT                                       Type: ELEMENT
    Value: null                                         Value: null
    Index: 1                                            Index: 1
    nMD: d125f297517596f8db901f0df9e65650               nMD: d125f297517596f8db901f0df9e65650
    tMD: c4e4c14e88367d1a00d323facd5a2e4d               tMD: ffc37f0ae203e623ce2bdd26d58980ef

        Operation: 0                                        Operation: 0
        Label: H1                                           Label: H1
        Type: ELEMENT                                       Type: ELEMENT
        Value: null                                         Value: null
        Index: 1                                            Index: 1
        nMD: b21f155a19cfc66ae9f794acd8dfea7d               nMD: b21f155a19cfc66ae9f794acd8dfea7d
        tMD: 7903405e2470de2b469ed7da0d80039f               tMD: 7903405e2470de2b469ed7da0d80039f

            Operation: 0                                        Operation: 0
            Label: #TEXT                                        Label: #TEXT
            Type: #TEXT                                         Type: #TEXT
            Value: hey!                                         Value: hey!
            Index: 1                                            Index: 1
            nMD: b158d340e3ff84a421d58f86c8d88ab1               nMD: b158d340e3ff84a421d58f86c8d88ab1
            tMD: fc67236df17dd901394e3bf2955309a8               tMD: fc67236df17dd901394e3bf2955309a8

        Operation: 3                                        Operation: 3
        Label: H1                                           Label: H1
        Type: ELEMENT                                       Type: ELEMENT
        Value: null                                         Value: null
        Index: 2                                            Index: 2
        nMD: b21f155a19cfc66ae9f794acd8dfea7d               nMD: b21f155a19cfc66ae9f794acd8dfea7d
        tMD: 7903405e2470de2b469ed7da0d80039f               tMD: 7622548960037d20e5325724349954af

            Operation: 3                                        Operation: 3
            Label: #TEXT                                        Label: #TEXT
            Type: #TEXT                                         Type: #TEXT
            Value: hey!                                         Value: well done!
            Index: 1                                            Index: 1
            nMD: b158d340e3ff84a421d58f86c8d88ab1               nMD: 90744f0f9d9ae8022bbc05d62dec3693
            tMD: fc67236df17dd901394e3bf2955309a8               tMD: 9e742919067e4a77adc5bd6cda5f1ef9

    */








    //           <div>
    //                  <h1>hey!</h1>
    //                  <h1>well done!</h1>
    //           </div>
    tree1 = DIV ( [
                      H1 (
                          [ hey () ] ,
                          1
                      ) ,
                      H1 (
                          [ well_done () ] ,
                          2
                      ) ,
                  ] )

    //           <div>
    //                  <h1>hey!</h1>
    //                  <h1>great work!</h1>
    //           </div>
    tree2 = DIV ( [
                      H1 (
                          [ hey () ] ,
                          1
                      ) ,
                      H1 (
                          [ great_work () ] ,
                          2
                      ) ,
                  ] )


    new DefaultXTreeDiff (
        tree1 ,
        tree2
    ).diff ()

    console.log ( getLog ( tree1 ) )
    console.log ( getLog ( tree2 ) )

    /*
    TREE 1                                            TREE 2
    ---------------------------------------           ---------------------------------------    
    Operation: 0                                      Operation: 0      
    Label: DIV                                        Label: DIV      
    Type: ELEMENT                                     Type: ELEMENT      
    Value: null                                       Value: null      
    Index: 1                                          Index: 1      
    nMD: d125f297517596f8db901f0df9e65650             nMD: d125f297517596f8db901f0df9e65650      
    tMD: ffc37f0ae203e623ce2bdd26d58980ef             tMD: d599210605ea74672c7c861d166357ad      

        Operation: 0                                      Operation: 0      
        Label: H1                                         Label: H1      
        Type: ELEMENT                                     Type: ELEMENT      
        Value: null                                       Value: null      
        Index: 1                                          Index: 1      
        nMD: b21f155a19cfc66ae9f794acd8dfea7d             nMD: b21f155a19cfc66ae9f794acd8dfea7d      
        tMD: 7903405e2470de2b469ed7da0d80039f             tMD: 7903405e2470de2b469ed7da0d80039f      

            Operation: 0                                      Operation: 0      
            Label: #TEXT                                      Label: #TEXT      
            Type: #TEXT                                       Type: #TEXT      
            Value: hey!                                       Value: hey!      
            Index: 1                                          Index: 1      
            nMD: b158d340e3ff84a421d58f86c8d88ab1             nMD: b158d340e3ff84a421d58f86c8d88ab1      
            tMD: fc67236df17dd901394e3bf2955309a8             tMD: fc67236df17dd901394e3bf2955309a8      

        Operation: 0                                      Operation: 0      
        Label: H1                                         Label: H1      
        Type: ELEMENT                                     Type: ELEMENT      
        Value: null                                       Value: null      
        Index: 2                                          Index: 2      
        nMD: b21f155a19cfc66ae9f794acd8dfea7d             nMD: b21f155a19cfc66ae9f794acd8dfea7d      
        tMD: 7622548960037d20e5325724349954af             tMD: 91e6319ebf6ac54bf6f808f52376e9f7      

            Operation: 0                                      Operation: 0      
            Label: #TEXT                                      Label: #TEXT      
            Type: #TEXT                                       Type: #TEXT      
            Value: well done!                                 Value: great work!      
            Index: 1                                          Index: 1      
            nMD: 90744f0f9d9ae8022bbc05d62dec3693             nMD: 67665e3c4f058fb5e33f487a9b78b0d7      
            tMD: 9e742919067e4a77adc5bd6cda5f1ef9             tMD: 060a16a1de8099c2b6833e8e67176459      

*/



```

**Comment by yidafu** [comment]
Hi, thanks for your report.

this because `X-Tree Diff+` doesn't cover this case.

![match remaining nodes downwards](https://user-images.githubusercontent.com/22773923/79961680-a331fd80-84b9-11ea-9fcd-c876d666adee.jpg)

When  `match remaining nodes downwards`, it only judged `label` and `index`. so Text Node `Op` 
 update incorrectly.

**Comment by yidafu** [comment]
Fix at [` @dovyih/x-tree-diff-plus@0.1.1`](https://www.npmjs.com/package/@dovyih/x-tree-diff-plus/v/0.1.1) @FUUbi

**Comment by FUUbi** [comment]
Wow, that was quick. Thanks, works like a charm.
